sled-agent/src/bootstrap/pre_server.rs

```diff
@@ -0,0 +1,536 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Functionality required to initialize the bootstrap agent even before it
+//! starts running servers on the bootstrap network.
+
+// Clippy doesn't like `StartError` due to
+// https://github.com/oxidecomputer/usdt/issues/133; remove this once that issue
+// is addressed.
+#![allow(clippy::result_large_err)]
+
+use super::maghemite;
+use super::secret_retriever::LrtqOrHardcodedSecretRetriever;
+use super::server::StartError;
+use crate::config::Config;
+use crate::services::ServiceManager;
+use crate::storage_manager::StorageManager;
+use camino::Utf8PathBuf;
+use cancel_safe_futures::TryStreamExt;
+use ddm_admin_client::Client as DdmAdminClient;
+use futures::stream;
+use futures::StreamExt;
+use illumos_utils::addrobj::AddrObject;
+use illumos_utils::dladm;
+use illumos_utils::dladm::Dladm;
+use illumos_utils::zfs;
+use illumos_utils::zfs::Zfs;
+use illumos_utils::zone;
+use illumos_utils::zone::Zones;
+use key_manager::KeyManager;
+use key_manager::StorageKeyRequester;
+use omicron_common::address::Ipv6Subnet;
+use omicron_common::api::internal::shared::RackNetworkConfig;
+use omicron_common::FileKv;
+use sled_hardware::underlay;
+use sled_hardware::DendriteAsic;
+use sled_hardware::HardwareManager;
+use sled_hardware::HardwareUpdate;
+use sled_hardware::SledMode;
+use slog::Drain;
+use slog::Logger;
+use std::net::IpAddr;
+use std::net::Ipv6Addr;
+use tokio::sync::broadcast;
+use tokio::task::JoinHandle;
+
+pub(super) struct BootstrapManagers {
+    pub(super) hardware: HardwareManager,
+    pub(super) storage: StorageManager,
+    pub(super) service: ServiceManager,
+}
+
+impl BootstrapManagers {
+    pub(super) async fn handle_hardware_update(
+        &self,
+        update: Result<HardwareUpdate, broadcast::error::RecvError>,
+        underlay_network: Option<(Ipv6Addr, Option<&RackNetworkConfig>)>,
+        log: &Logger,
+    ) {
+        match update {
+            Ok(update) => match update {
+                HardwareUpdate::TofinoLoaded => {
+                    let baseboard = self.hardware.baseboard();
+                    if let Err(e) = self
+                        .service
+                        .activate_switch(underlay_network, baseboard)
+                        .await
+                    {
+                        warn!(log, "Failed to activate switch: {e}");
+                    }
+                }
+                HardwareUpdate::TofinoUnloaded => {
+                    if let Err(e) = self.service.deactivate_switch().await {
+                        warn!(log, "Failed to deactivate switch: {e}");
+                    }
+                }
+                HardwareUpdate::TofinoDeviceChange => {
+                    // TODO-correctness What should we do here?
+                }
```

Comment:
Makes sense. I cleaned this up a bit (and made it more like the previous sled-agent code) in 391b29dd: instead of taking only the underlay info, we now take an `Option<&SledAgent>`. With that we can get the underlay info (if it exists) and also directly call `notify_nexus_about_self()` here; we also to remove the very awkward `should_notify_nexus` stuff in our caller.